towerlib/towerlib.py

```diff
@@ -2018,3 +2034,300 @@ def get_schedule_by_name(self, name):
+        """Get job creation dates from ansible tower for a given host
+
+        Args:
+            host: the host object of tower instance.
+
+        Returns:
+            list: a list containing the formatted datetime objects.
+
+        """
+        job_events = list(host.job_events)
+        if job_events is None:
+            self._logger.error("No job events found.")
+            return None
+        return [job_event.created_at for job_event in job_events]
+
+    def get_groups_by_host(self, host):
+        """Get groups for a particular host, which are directly connected.
+
+        Args:
+            host: the host object.
+
+        Returns:
+            list: list of custom groups.
+
+        """
+        url = "{api}/hosts/{id}/groups/".format(api=self.api, id=host.id)
+        response = self.session.get(url)
+        if not response.ok:
+            self._logger.error("Error getting the project updates. response was: {})".format(response.text))
+            return None
+        else:
+            results = response.json().get('results', [])
+            groups = []
+            for group in results:
+                groups.append(group)
+            return groups
+
+    def get_all_groups_by_host(self, host):
+        """Get groups for a particular host, which are directly and indirectly connected.
+
+        Args:
+            host: the host object.
+
+        Returns:
+            list: list of custom groups.
+
+        """
+        url = "{api}/hosts/{id}/all_groups/".format(api=self.api, id=host.id)
+        response = self.session.get(url)
+        if not response.ok:
+            self._logger.error("Error getting the project updates. response was: {})".format(response.text))
+            return None
+        else:
+            results = response.json().get('results', [])
+            groups = []
+            for group in results:
+                groups.append(group)
+            return groups
+
+    def search_generic_item_by_keyword(self, generic_item_name_plural, keyword=''):
+        """
+        Search query string parameter to perform a case-insensitive search within all designated text fields of a model
+        or item. Model/item means projects, jobs, inventories etc.
+        https://docs.ansible.com/ansible-tower/latest/html/towerapi/searching.html
+
+        Args:
+            generic_item_name_plural: the plural name of the model/item e.g. projects, jobs, job_templates, users etc.
+            keyword: case insensitive search string.
+
+        Returns:
+            list: list of matching item objects.
+
+        """
+        if not generic_item_name_plural.lower() in GENERIC_SEARCH_ITEMS:
+            self._logger.error("The generic search item '{}' not found.".format(generic_item_name_plural))
+            return None
+        url = "{api}/{item}/?search={keyword}".format(api=self.api, item=generic_item_name_plural, keyword=keyword)
+        response = self.session.get(url)
```

Comment:
There is a method `def _get_paginated_response(self, url, params=None):` that implements pagination and should be used instead of reimplementing the logic.

Reply:
Could u please help me refactor this method? I have tried the way in attached screenshot: 
![image](https://user-images.githubusercontent.com/18288421/123221026-cb0e5e00-d4ce-11eb-8ff6-f62fc3549f21.png)

But this only giving me partial results. For example, if there are total 474 job_templates, I get only 12 job_templates if I look for job_templates with this search method.

Reply:
Might be a dumb question but what would the output be if you try out the query from the api itself? Could it be that the search you are doing actually filters down the output?

Reply:
I have tried the query directly from the api e.g. _{host}/api/v2/job_templates/?search=_ 
and I got all the job_templates (474). Did you mean something like this?